_posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd

```diff
@@ -0,0 +1,326 @@
+
+But how does this translate into neural network operations? It's just another matrix multiplication ($V^T \mathbf{h}$) added
+to the convolutional outputs ($W \mathbf{x}$).
+
+$$\mathbf{y} = tanh(W_{k,f} \mathbf{x} + V^T_{k,f} \mathbf{h}) \odot \sigma(W_{k,g} \mathbf{x} + V^T_{k,g} \mathbf{h})$$
+
+(If you're wondering about the second part on the right, after the Hadamard product sign -- we won't go into details, but in a
+nutshell, it's another modification introduced by [@OordKVEGK16], a transfer of the "gating" principle from recurrent neural
+networks, such as GRUs and LSTMs, to the convolutional setting.)
+
+So we see what goes into the decision of a pixel value to sample. But how is that decision actually *made*?
+
+### Logistic mixture likelihood , or: No pixel is an island
+
+Again, this is where the TFP implementation does not follow the original paper, but the latter PixelCNN++ one. Originally,
+pixels were modeled as discrete values, decided on by a softmax over 256 (0-255) possible values. (That this actually worked
+seems like another instance of deep learning magic. Imagine: In this model, 254 is as far from 255 as it is from 0.)
+
+In contrast, PixelCNN++ assumes an underlying continuous distribution of color intensity, and rounds to the nearest integer.
+That underlying distribution is a mixture of logistic distributions, thus allowing for multimodality:
+
+$$\nu \sim \sum_{i} \pi_i \ logistic(\mu_i, \sigma_i)$$
+
+### Overall architecture and the PixelCNN distribution
+
+Overall, PixelCNN++, as described in [@Salimans2017PixeCNN], consists of six blocks. The blocks together make up a UNet-like
+structure, successively downsizing the input and then, upsampling again:
+
+![Fig. 2: Overall structure of PixelCNN++. From: Salimans et al., 2017.](images/high-level.png)
+
+In TFP's PixelCNN distribution, the number of blocks is configurable as `num_hierarchies`, the default being 3.
+
+Each block consists of a customizable number of layers, called *ResNet layers* due to the residual connection (visible on the
+right) complementing the convolutional operations in the horizontal stack:
+
+![Fig. 3: One so-called \"ResNet layer\", featuring both a vertical and a horizontal convolutional stack. Source: van den Oord
+et al., 2017.](images/layer.png)
+
+In TFP, the number of these layers per block is configurable as `num_resnet`.
+
+`num_resnet` and `num_hierarchies` are the parameters you're most likely to experiment with, but there are a few more you can
+check out in the [documentation](https://rstudio.github.io/tfprobability/reference/tfd_pixel_cnn.html). The number of logistic
+distributions in the mixture is also configurable, but from my experiments it's best to keep that number rather low to avoid
+producing `NaN`s during training.
+
+Let's now see a complete example.
+
+## End-to-end example
+
+Our playground will be [QuickDraw](tbd), a dataset -- still growing -- obtained by asking people to draw some object in at
```

Comment:
Just a reminder that you've got the ***tbd*** for the link here!

Reply:
ooops ... thanks!